synapse_data_warehouse/synapse/dynamic_tables/R__populate_acl_latest_table.sql

```diff
@@ -0,0 +1,57 @@
+-- 1. Initialize the schema that the dynamic table will live under
+USE SCHEMA {{database_name}}.synapse; --noqa: JJ01,PRS,TMP,CP01
+
+-- 2. Create the first temporary table which unpacks the resource_access column
+--    and extracts the access type and principal ID as new columns
+CREATE TEMPORARY TABLE acl_expanded_temp AS
```

Comment:
Just a note: the temporary tables need to be within the dynamic table statement, or else they won't be created every single time.

Reply:
Ah yeah. Since the purpose of using the temporary tables was to simplify the dynamic table logic, I don't think it's helpful to use temporary/transient tables in this case. What I can do to help with code clarity in the dynamic table logic is add more comments to separate the CTEs to provide more structure.

What do you think of going in that direction @philerooski ?

Reply:
Could you explain what you mean, @thomasyu888 ? How can the dynamic table reference the temporary table `dedup_acl_expanded_temp` without that table being created as part of the session?

My intention with factoring out the CTEs was to reduce the amount of nesting in the CREATE TABLE statement. We would never nest functions three-levels deep across 45 lines in _a single statement_ in Python or R, so why do it in SQL if we don't have to?

Reply:
@philerooski So the raw tables are updated with data, so you would have to ensure these temporary tables are created every time the dynamic table queries are run.  Temporary tables contain static information, this particular SQL code isn't run on a schedule, but the code defined in the dynamic tables code is.

Separately, CTE's are meant to serve as "functions", so in my opinion, the above is technically proper usage of CTEs.  now I could eventually see the separation of bronze, silver, gold tables, where there is a middle step of (silver tables) containing the first layer, then the gold tables containing the last.

Reply:
I was going to suggest using `TARGET_LAG` on the temporary tables, but this seems to be a feature unique to dynamic tables. From what I'm reading, it looks like CTEs are the best use-case for this type of thing, since all the logic that runs on a schedule based on `TARGET_LAG` is under the `CREATE OR REPLACE DYNAMIC TABLE...` logic. 

I really prefer the script the way it's structured with the temporary tables, so I'll keep poking around the docs to see if there's an alternative, but for now what do we think of us going forward with the CTEs and structuring them with comments?

Reply:
@thomasyu888 That makes sense. There's a huge list of [limitations](https://docs.snowflake.com/en/user-guide/dynamic-tables-limitations) that come with using dynamic tables, and weirdly enough temporary tables aren't mentioned except indirectly: "You can’t use dynamic SQL (for example, session variables or unbound variables of anonymous blocks) in the dynamic table’s definition."

Probably it's not worth trying to avoid these behemoth dynamic table definitions.